### ISSUE TITLE
Image Service doesn't provide CORS header

### ISSUE THREAD

**Issue by warthds2** [storyblok/storyblok#580]
**The issue occurs while working with:** (check one with "x")   
- [ ] *app.storyblok.com (CMS - Interface)*   
- [ ] *api.storyblok.com (CMS - Content Delivery API)*
- [ ] *mapi.storyblok.com (CMS - Management API)*
- [ ] *capi.storyblok.com (Commerce - API)*     
- [ ] *Commerce - Interface*   
- [X] *Other:* https://img2.storyblok.com/

**Current behavior:**
The images delivered through Image Service doesn't provide CORS header (access-control-allow-origin):

![Screenshot 2021-01-01 at 15 49 23](https://user-images.githubusercontent.com/50714010/103440847-44787280-4c49-11eb-8b93-a99b1c590f61.png)

**Expected behavior:**
The response provides CORS header.

**Steps to reproduce:**
Check request and response of the provided URL: https://img2.storyblok.com/fit-in/200x200/filters:fill(transparent):format(png)/f/39898/3310x2192/e4ec08624e/demo-image.jpeg

**Other information:**
I saw this issue: https://github.com/storyblok/storyblok/issues/380.

I checked the suggestion to request the image by the s3 AWS url. Unfortunately it doesn't work: 

https://s3.amazonaws.com/img2.storyblok.com/fit-in/200x200/filters:fill(transparent):format(png)/f/39898/3310x2192/e4ec08624e/demo-image.jpeg

<Error>
<Code>NoSuchBucket</Code>
<Message>The specified bucket does not exist</Message>
<BucketName>img2.storyblok.com</BucketName>
<RequestId>EV9Q4G0Q8G9TFZAW</RequestId>
<HostId>fpKmXdj1TvUdONYhBKJryEWZb5gYFDccRD7bJ+uUKhDTccCmXIUW19/G0Kl/5la5468yWE6fKXk=</HostId>
</Error>

**Comment by sever1an** [comment]
This is a major issue for us as we are migrating to using a service worker to implement a progressive web app. Opaque responses are extremely difficult to cache, due to having no insight into whether the request was successful, and that browsers hide the actual size of the files by allocating an **enormous** amount of storage from the quota. This is on the order of hundreds of mb for a one mb file. 

At this point our PWA project on a storyblok site is dead. **Our only option moving forward is to select a different CMS product.** Is there any intention to allow CORS access to assets?

**Comment by sever1an** [comment]
Further research suggests that proxying the source image urls through imgix might be an alternative solution. We would rather just use the resize service built into Storyblok though!

**Comment by sever1an** [comment]
I am actually rather enamoured with imgix and withdraw my original complaint.

**Comment by dohomi** [comment]
I have this requirement as well and don't quite understand why a.storyblok and img2.storyblok doesn't set the cors headers. Is there any way to get around this issue? Will the endpoints getting an update?

**Comment by jackismissing** [comment]
Having the same issue here as well, will we have any update on this? Not being able to access images in Javascript because headers can't be set on the a.storyblok endpoint is quite a problem for us.

**Comment by BURROO** [comment]
I am also running into the same issue. Anyone found a solution meanwhile?

**Comment by jackismissing** [comment]
More than a year later this issue still hasn't been addressed... I tried the "new" API and even there CORS headers won't be set properly:
storyblok.com/docs/image-service

**Comment by victorronnow** [comment]
Is there any updates on this? I understand fixing issues take time and I'm willing to be patient, but this is literally one if not the only thing that prevents me from using Storyblok at this time. I really hope you can provide us with an ETA on when this is getting fixed.

**Comment by mwmwmw** [comment]
CORS on assets is a must, especially if you want to support highly interactive sites that use WebGL and other media because they're so tight on CORS. 

Could the many CORS issues be related to local development?

**Comment by michaelpumo** [comment]
I was hoping to get the "dominant" colour from an image using canvas from a Storyblok image but CORS is preventing this. The only other alternative is to proxy the URL through my own domain but that requires server functions and I was hoping to keep my app static. Ahh.

**Comment by victorronnow** [comment]
After chatting with the support team at Storyblok I finally have a solution. For anyone having this issue using `a2.storyblok.com` instead of `a.storyblok.com` in the asset url provides CORS. It worked for me at least. It also works with Image Service.

**Comment by michaelpumo** [comment]
> 



> After chatting with the support team at Storyblok, I finally have a solution. For anyone having this issue using `a2.storyblok.com` instead of `a.storyblok.com` in the asset url provides CORS. It worked for me at least. It also works with Image Service. I don't know why CORS aren't enabled by default.
> 
> I know this is suboptimal, but I use this JavaScript function to replace the url:
> 
> ```
> const getUrl = (url) => url.replace('//a.storyblok.com', '//a2.storyblok.com');
> ```

Wow, this way worked perfectly. Thank you.

**Comment by mwmwmw** [comment]
> const getUrl = (url) => url.replace('//a.storyblok.com', '//a2.storyblok.com');

This worked for me as well, but I had a regional URL.

I did this

```
filename.replace('//a-us.storyblok.com', '//a2-us.storyblok.com');
```

Works perfectly.

**Comment by weicheng-one** [comment]
> > 
> 
> > After chatting with the support team at Storyblok, I finally have a solution. For anyone having this issue using `a2.storyblok.com` instead of `a.storyblok.com` in the asset url provides CORS. It worked for me at least. It also works with Image Service. I don't know why CORS aren't enabled by default.
> > I know this is suboptimal, but I use this JavaScript function to replace the url:
> > ```
> > const getUrl = (url) => url.replace('//a.storyblok.com', '//a2.storyblok.com');
> > ```
> 
> Wow, this way worked perfectly. Thank you.

Thanks, It's work perfectly!

**Comment by fabiandarga** [comment]
That can not be the permanent solution. 
I was hoping I can host some json or csv files in assets to be consumed by a small script on our website.
Weirdly it worked for a day, now I get CORS problems. We did not change anything in between.

**Comment by nephix** [comment]
Running into the same issue when adding subtitles (`.vtt` files) to the asset library and using them in bloks

**Comment by awacode21** [comment]
The issue came up in 2021, now we have 2023 without any progress. Yes there is a string replace workaround, but honestly... this should not be the solution. The images are hosted with storyblok. So handling CORS is a must!!!! We should not wait over 2 years for a fix.

**Comment by SimonKib** [comment]
The very least, Storyblok team should mention this option on https://www.storyblok.com/docs/image-service/. I spent too much time looking for other workarounds than `replace('https://a', 'https://a2') `

**Comment by Kethatril** [comment]
Just ran into this again, and sadly the a2 urls don't seem to exist for every region, specifically https://a2-ap.storyblok.com isn't a thing.

Note that https://a.storyblok.com/ does have CORS headers **only** if you access it first via javascript when loading the image to use in webgl. It seems to then cache those headers for all future responses to that asset. However if you access it directly via a URL, or via a normal image element then it doesn't set any CORS headers and caches that lack of headers.

**Comment by codeflorist** [comment]
Check out this FAQ:
https://www.storyblok.com/faq/xmlhttprequest-cannot-load-no-access-control-allow-origin-header

It suggests using using `https://s3.amazonaws.com/a.storyblok.com` instead of `https://a.storyblok.com`.

**Comment by Noaber** [comment]
Just to be sure, the current "workaround" is replacing a.storyblok with a2.storyblok ?

**Comment by AdrianFahrbach** [comment]
Yes, it seems like this is the only current workaround. We did test the solution mentioned in the docs [here](https://www.storyblok.com/faq/xmlhttprequest-cannot-load-no-access-control-allow-origin-header) first, but this didn't work.

@alvarosabu, @christianzoppi, @manuelschroederdev, @iamfortune, @samuells could anyone of you check this and initiate a corresponding change in the documentation? It seems like this is a "solved" issue.

**Comment by alvarosabu** [comment]
Hi @warthds2, we are closing this issue because this repository was used in the early days for support-related questions, but we've since transitioned to more effective channels for helping our users.

Please, refer to the following resources:

Storyblok Github Org: https://github.com/storyblok
Storyblok Docs: https://www.storyblok.com/docs
Official Support: https://support.storyblok.com
Join our Discord: https://www.storyblok.com/join-discord

Thanks for the understanding 🙏🏻
